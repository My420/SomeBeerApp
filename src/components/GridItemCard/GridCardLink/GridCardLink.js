import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import GridCardDescription from '../GridCardDescription/GridCardDescription';
import styles from './GridCardLink.module.scss';
import animation from './animation.module.css';
import replaceEmptyImageUrl from '../../../utils/replaceEmptyImageUrl';

const GridCardLink = ({
  data,
  isDescriptionOpen,
  onCursorEnter,
  onCursorLeave
}) => {
  const parameter = {
    abv: data.abv,
    ibu: data.ibu,
    ebc: data.ebc
  };

  const addEndListener = (node, done) => {
    node.addEventListener('transitionend', done, false);
  };

  // eslint-disable-next-line no-console
  console.log('render ===== GridCardLink');

  return (
    <Link
      className={styles.link}
      to={`id/${data.id}`}
      onMouseEnter={onCursorEnter}
      onMouseLeave={onCursorLeave}
    >
      <div className={styles.wrapper}>
        <TransitionGroup component={null}>
          {isDescriptionOpen ? (
            <CSSTransition
              classNames={{ ...animation }}
              timeout={200}
              addEndListener={addEndListener}
            >
              <GridCardDescription parameter={parameter} title={data.tagline} />
            </CSSTransition>
          ) : null}
        </TransitionGroup>
        <img
          className={styles.image}
          src={replaceEmptyImageUrl(data.image_url)}
          alt={data.name}
        />
      </div>
      <p className={styles.description}>
        <span className={styles.beerName}>{data.name}</span>
      </p>
    </Link>
  );
};

GridCardLink.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  data: PropTypes.object.isRequired,
  isDescriptionOpen: PropTypes.bool.isRequired,
  onCursorEnter: PropTypes.func.isRequired,
  onCursorLeave: PropTypes.func.isRequired
};

export default GridCardLink;
