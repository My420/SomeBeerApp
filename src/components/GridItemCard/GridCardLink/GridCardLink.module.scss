@import '../../../sass/variables.scss';

.link {
  display: block;

  margin: 0 auto;

  width: 266px;
  height: 276px;

  overflow: hidden;

  text-align: center;
  text-decoration: none;

  color: $menu-color;
}

.link:focus {
  outline: 2px solid $menu-hover-color;
}

.wrapper {
  display: flex;
  justify-content: center;

  position: relative;

  width: 100%;
  height: 200px;

  padding-top: 10px;
}

.image {
  max-width: 100%;
  max-height: 100%;
}

.description {
  display: flex;
  justify-content: center;
  align-items: center;

  height: 50px;
  padding: 10px 5px;
}

.beerName {
  max-height: 50px;
  overflow: hidden;
}
