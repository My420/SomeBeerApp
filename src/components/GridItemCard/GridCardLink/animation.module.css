.appear {
}
.appearActive {
}
.appearDone {
}
.enter {
  transform: translateY(-100%);
}
.enterActive {
  transform: translateY(0);
  transition: all 200ms linear;
}
.enterDone {
}
.exit {
  transform: translateY(0);
}
.exitActive {
  transform: translateY(-100%);
  transition: all 200ms linear;
}
.exitDone {
}
