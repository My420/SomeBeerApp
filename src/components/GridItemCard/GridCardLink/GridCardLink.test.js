import React from 'react';
import { shallow } from 'enzyme';
import GridCardLink from './GridCardLink';

describe('test <GridCardLink/> component', () => {
  const data = {
    abv: 11,
    ibu: 22,
    ebc: 33,
    id: 11,
    tagline: 'test tagline',
    image_url: 'test url',
    name: 'test name'
  };

  const enterMock = jest.fn();
  const leaveMock = jest.fn();

  describe('test component when prop isDescriptionOpen = false', () => {
    const wrapper = shallow(
      <GridCardLink
        data={data}
        isDescriptionOpen={false}
        onCursorEnter={enterMock}
        onCursorLeave={leaveMock}
      />
    );

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test("should render Link component with correct prop 'to'", () => {
      const component = wrapper.find('Link');
      expect(component).toHaveLength(1);
      expect(component.prop('to')).toBe(`id/${data.id}`);
    });

    test('should render tag <img/> with correct attr', () => {
      const component = wrapper.find('img');
      expect(component).toHaveLength(1);
      expect(component.prop('alt')).toBe(data.name);
      expect(component.prop('src')).toBe(data.image_url);
    });

    test('should render tag <span/> with correct text', () => {
      const component = wrapper.find('span');
      expect(component).toHaveLength(1);
      expect(component.text()).toBe(data.name);
    });
  });

  describe('test component when prop isDescriptionOpen = true', () => {
    const wrapper = shallow(
      <GridCardLink
        data={data}
        isDescriptionOpen
        onCursorEnter={enterMock}
        onCursorLeave={leaveMock}
      />
    );

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test("should render Link component with correct prop 'to'", () => {
      const component = wrapper.find('Link');
      expect(component).toHaveLength(1);
      expect(component.prop('to')).toBe(`id/${data.id}`);
    });

    test('should render tag <img/> with correct attr', () => {
      const component = wrapper.find('img');
      expect(component).toHaveLength(1);
      expect(component.prop('alt')).toBe(data.name);
      expect(component.prop('src')).toBe(data.image_url);
    });

    test('should render tag <span/> with correct text', () => {
      const component = wrapper.find('span');
      expect(component).toHaveLength(1);
      expect(component.text()).toBe(data.name);
    });

    test('should render GridCardDescription component with correct props', () => {
      const { abv, ebc, ibu, tagline } = data;
      const expectedParam = {
        abv,
        ebc,
        ibu
      };
      const component = wrapper.find('GridCardDescription');
      expect(component).toHaveLength(1);
      expect(component.prop('parameter')).toEqual(expectedParam);
      expect(component.prop('title')).toBe(tagline);
    });
  });

  describe('test component mouseEnter and mouseLeave event', () => {
    const wrapper = shallow(
      <GridCardLink
        data={data}
        isDescriptionOpen={false}
        onCursorEnter={enterMock}
        onCursorLeave={leaveMock}
      />
    );

    test('should call onCursorEnter prop when mouse above component', () => {
      const component = wrapper.find('Link');
      component.simulate('mouseEnter');
      expect(enterMock.mock.calls.length).toBe(1);
    });

    test('should call onCursorLeave prop when mouse leave component', () => {
      const component = wrapper.find('Link');
      component.simulate('mouseLeave');
      expect(leaveMock.mock.calls.length).toBe(1);
    });
  });
});
