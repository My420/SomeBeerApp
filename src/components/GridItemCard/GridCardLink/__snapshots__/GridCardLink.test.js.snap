// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <GridCardLink/> component test component when prop isDescriptionOpen = false should match snapshot 1`] = `
<Link
  className="link"
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
  to="id/11"
>
  <div
    className="wrapper"
  >
    <TransitionGroup
      childFactory={[Function]}
      component={null}
    />
    <img
      alt="test name"
      className="image"
      src="test url"
    />
  </div>
  <p
    className="description"
  >
    <span
      className="beerName"
    >
      test name
    </span>
  </p>
</Link>
`;

exports[`test <GridCardLink/> component test component when prop isDescriptionOpen = true should match snapshot 1`] = `
<Link
  className="link"
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
  to="id/11"
>
  <div
    className="wrapper"
  >
    <TransitionGroup
      childFactory={[Function]}
      component={null}
    >
      <CSSTransition
        addEndListener={[Function]}
        classNames={Object {}}
        timeout={200}
      >
        <GridCardDescription
          parameter={
            Object {
              "abv": 11,
              "ebc": 33,
              "ibu": 22,
            }
          }
          title="test tagline"
        />
      </CSSTransition>
    </TransitionGroup>
    <img
      alt="test name"
      className="image"
      src="test url"
    />
  </div>
  <p
    className="description"
  >
    <span
      className="beerName"
    >
      test name
    </span>
  </p>
</Link>
`;
